hibernate-core/src/main/java/org/hibernate/sql/model/ast/builder/CollectionRowUpdateBuilder.java

```diff
@@ -0,0 +1,87 @@
+/*
+ * Hibernate, Relational Persistence for Idiomatic Java
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later
+ * See the lgpl.txt file in the root directory or http://www.gnu.org/licenses/lgpl-2.1.html
+ */
+package org.hibernate.sql.model.ast.builder;
+
+import java.util.List;
+
+import org.hibernate.engine.spi.SessionFactoryImplementor;
+import org.hibernate.jdbc.Expectation;
+import org.hibernate.persister.collection.mutation.CollectionTableMapping;
+import org.hibernate.sql.model.MutationOperation;
+import org.hibernate.sql.model.MutationTarget;
+import org.hibernate.sql.model.ast.ColumnValueBinding;
+import org.hibernate.sql.model.ast.MutatingTableReference;
+import org.hibernate.sql.model.ast.RestrictedTableMutation;
+import org.hibernate.sql.model.internal.TableUpdateCustomSql;
+import org.hibernate.sql.model.internal.TableUpdateStandard;
+
+/**
+ * Custom table update builder for one-to-many collections that handles row deletes
+ *
+ * @author Marco Belladelli
+ */
+public class CollectionRowUpdateBuilder<O extends MutationOperation> extends TableUpdateBuilderStandard<O> {
+	public CollectionRowUpdateBuilder(
+			MutationTarget<?> mutationTarget,
+			MutatingTableReference tableReference,
+			SessionFactoryImplementor sessionFactory,
+			String whereFragment) {
+		super( mutationTarget, tableReference, sessionFactory, whereFragment );
+		assert tableReference.getTableMapping() instanceof CollectionTableMapping;
+	}
+
+	@SuppressWarnings( "unchecked" )
+	@Override
+	public RestrictedTableMutation<O> buildMutation() {
+		final CollectionTableMapping tableMapping = (CollectionTableMapping) getMutatingTable().getTableMapping();
+		final List<ColumnValueBinding> valueBindings = combine(
+				getValueBindings(),
+				getKeyBindings(),
+				getLobValueBindings()
+		);
+		if ( tableMapping.getDeleteRowDetails().getCustomSql() != null ) {
+			return (RestrictedTableMutation<O>) new TableUpdateCustomSql(
+					getMutatingTable(),
+					getMutationTarget(),
+					getSqlComment(),
+					valueBindings,
+					getKeyRestrictionBindings(),
+					getOptimisticLockBindings()
+			) {
+				@Override
+				public String getCustomSql() {
+					return tableMapping.getDeleteRowDetails().getCustomSql();
+				}
+
+				@Override
+				public boolean isCallable() {
+					return tableMapping.getDeleteRowDetails().isCallable();
+				}
+
+				@Override
+				public Expectation getExpectation() {
+					return tableMapping.getDeleteRowDetails().getExpectation();
+				}
+			};
+		}
+		return (RestrictedTableMutation<O>) new TableUpdateStandard(
+				getMutatingTable(),
+				getMutationTarget(),
+				getSqlComment(),
+				valueBindings,
+				getKeyRestrictionBindings(),
+				getOptimisticLockBindings(),
+				getWhereFragment(),
+				null
+		) {
+			@Override
+			public Expectation getExpectation() {
+				return tableMapping.getDeleteRowDetails().getExpectation();
```

Comment:
```suggestion
				return tableMapping.getUpdateRowDetails().getExpectation();
```

Reply:
I've used `getDeleteRowDetails()` in `CollectionRowUpdateBuilder` because it is really used for **deleting** single row records for one-to-many associations, see `OneToManyPersister#generateDeleteRowAst`.

Maybe the naming I used for the builder is confusing and should be changed, but I don't think the suggested changes should be applied here.

Reply:
Indeed, then how about naming the class `CollectionRowDeleteByUpdateSetNullBuilder`?

Reply:
Yeah, that's a lot more expressive. Applied the new class name, thanks.